File: solutions/problem_1546_maximum_number_of_non_overlapping_subarrays_with_sum_equals_target/solution.py
"""
Problem 1546: Maximum Number of Non-Overlapping Subarrays With Sum Equals Target
==============================================================================
Difficulty: Medium
Tags: Array, Hash Table, Greedy, Prefix Sum
Pattern: Greedy

Time Complexity:  O(n log n)
Space Complexity: O(1)
"""

class Solution:
    def maxNonOverlapping(self, nums: List[int], target: int) -> int:
        # Greedy approach - O(n) time
        result = 0
        curr_max = 0
        for i in range(len(nums)):
            if isinstance(nums[i], int):
                curr_max = max(curr_max, nums[i])
                result = max(result, curr_max)
            else:
                result += 1
        return result

File: solutions/problem_1546_maximum_number_of_non_overlapping_subarrays_with_sum_equals_target/solution.cpp
/*
 * Problem 1546: Maximum Number of Non-Overlapping Subarrays With Sum Equals Target
 * ==============================================================================
 * Difficulty: Medium
 * Tags: Array, Hash Table, Greedy, Prefix Sum
 * Pattern: Greedy
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(1)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int maxNonOverlapping(vector<int>& nums, int target) {
        // Greedy approach - O(n) time
        int result = 0, curr_max = 0;
        for (int i = 0; i < (int)nums.size(); i++) {
            curr_max = max(curr_max, nums[i]);
            result = max(result, curr_max);
        }
        return result;
    }
};
